From 53348573fd5cc4522c844541c8488dd2ab1cdd4e Mon Sep 17 00:00:00 2001
From: Howard Lin <howard-yh.lin@mediatek.com>
Date: Tue, 24 Oct 2023 17:22:25 +0800
Subject: [PATCH] GENIO: cmd: add support for fdtprobe command

This allows adding 'fdtprobe' for probing & loading fdt files in
firmware dedicated partitions when 'secure-boot' is in DISTRO_FEATURES.

Change-Id: I92533f8edbf1828a3c88f1cec70aba79d1a8386d
Signed-off-by: Howard Lin <howard-yh.lin@mediatek.com>
---
 cmd/Kconfig    |  6 ++++++
 cmd/dtbprobe.c | 23 +++++++++++++++++++++++
 2 files changed, 29 insertions(+)

diff --git a/cmd/Kconfig b/cmd/Kconfig
index 0e0be94f41..869804ab09 100644
--- a/cmd/Kconfig
+++ b/cmd/Kconfig
@@ -526,6 +526,12 @@ config CMD_ZBOOT
 	  booting both 32- and 64-bit kernels, as well as secure boot.
 	  Documentation is available in doc/uImage.FIT/x86-fit-boot.txt
 
+config CMD_FDTPROBE
+	bool "Enable FDT probe command"
+	default n
+	help
+	  This provides probing and loading FDT files in firmware dedicated partitions.
+
 endmenu
 
 menu "Environment commands"
diff --git a/cmd/dtbprobe.c b/cmd/dtbprobe.c
index 3ef6464cf8..875c2caec7 100644
--- a/cmd/dtbprobe.c
+++ b/cmd/dtbprobe.c
@@ -90,6 +90,12 @@ static int loaddtb(char *devtype, char *devnum, int part, char *path, char *fdt_
 	if (run_command_list(cmd, -1, 0) != CMD_RET_SUCCESS)
 		return 1;
 
+#if defined(CONFIG_CMD_FDTPROBE)
+	sprintf(cmd, "fdt authndtb %s", fdt_addr);
+	if (run_command_list(cmd, -1, 0) != CMD_RET_SUCCESS)
+		return 1;
+#endif
+
 	sprintf(cmd, "fdt resize 8192");
 	if (run_command_list(cmd, -1, 0) != CMD_RET_SUCCESS)
 		return 1;
@@ -131,6 +137,14 @@ static int loadoverlays(char *devtype, char *devnum, int part, char *base)
 			goto out;
 		}
 
+#if defined(CONFIG_CMD_FDTPROBE)
+		sprintf(cmd, "fdt authndtb %s", overlay_addr);
+		if (run_command_list(cmd, -1, 0) != CMD_RET_SUCCESS) {
+			rcode = 1;
+			goto out;
+		}
+#endif
+
 		sprintf(cmd, "fdt apply %s", overlay_addr);
 		if (run_command_list(cmd, -1, 0) != CMD_RET_SUCCESS) {
 			rcode = 1;
@@ -210,9 +224,18 @@ static int do_dtbprobe(struct cmd_tbl *cmdtp, int flag, int argc,
 	return 1;
 }
 
+#if defined(CONFIG_CMD_FDTPROBE)
+U_BOOT_CMD(
+	fdtprobe,	CONFIG_SYS_MAXARGS,	1,	do_dtbprobe,
+	"Probe and load fdt from given partition list",
+	"<interface> <dev> <probe_base>\n"
+	"<probe_base>: The base directory for probing fdt files\n"
+);
+#else
 U_BOOT_CMD(
 	dtbprobe,	CONFIG_SYS_MAXARGS,	1,	do_dtbprobe,
 	"Probe and load dtb/dtbo from given partition list",
 	"<interface> <dev> <probe_base>\n"
 	"<probe_base>: The base directory for probing dtb files\n"
 );
+#endif
-- 
2.18.0

